docs/RFCS/20171003_inverted_indexes.md

```diff
@@ -0,0 +1,506 @@
+limit for RocksDB. The only way to get around that limit would be to add logic
+to split and merge values, adding unnecessary complexity.
+
+Fortunately CockroachDB has already solved this problem for non-unique
+secondary indexes where we encoding missing primary key components as part of
+the index key. Inverted indexes can reuse this approach to avoid the
+disadvantages of the posting list. We can encode both the full JSON path and
+the missing primary key components into the index key.
+
+As an implementation note, Postgres deviates from this idea by storing a hash
+of all of values of the path components as its index key to save space. This
+space savings comes at the cost of losing a lot of index capabilities: sub-path
+matching, top-level-key matching, and range scans of path and leaf values. This
+approach isn’t useful for us, since all of our keys are stored in sorted order
+and RocksDB does prefix compression on adjacent keys, so we don’t have to
+sacrifice those neat capabilities to save space.
+
+**Physical encoding**
+
+JSON values contain two collection types:
+
+- Objects: collections of key/value pairs
+- Arrays: Ordered lists of values
+
+We will encode these two structures into index keys in a very similar way. The
+full key schema looks like this:
+
+`/table id/index id/(json object key | arr)…/json leaf value/pk columns…/`
+
+where `arr` is a separator indicating that the next key is part of an array.
+There’s only one leaf value per index key, so the start of the primary key
+columns always follows the leaf value, which is detectable by its encoding. The
+value component of an index KV is left unused. In the future, inverted indexes
+could be extended by adding stored columns into the value component like
+secondary indexes currently allow with `STORING`.
+
+**Key encoding JSON objects**
+
+Encoding JSON objects into the index key is straightforward. Every key in a
+path is written to the index key after the index id and is prefixed with a type
+tag. Object keys don’t need ordinary type tags, because they can only be
+strings. However, we still need to add some sort of tag to distinguish them
+from leaf values, which will need to be encoded with a type tag. We’ll pick
+`BYTES`, which is not used by the leaf value encoding.
+
+JSON leaf values have four types, `null`, boolean, string or number. Since we
+don’t know what they are ahead of time, we’ll need to tag them with types.
+CockroachDB already uses type tags for its value encoding, so we can reuse
+those here.
+
+For example, if we had a table with id `1` with columns `keyCol` and `jsonCol`
+containing the following values:
+
+| keyCol | jsonCol                                    |
+| ------ | ------------------------------------------ |
+| `pk1`  | `{"x": "b", "z": {"a": true, "b": false}}` |
+| `pk2`  | `{"x": "b"}`                               |
+
+and an inverted index on `jsonCol` with index id `2`, we’d encode the following
+index keys:
+
+    1/2/x/b/pk1
```

Comment:
Do we also need to write keys for empty leaf values? For instance, do we need `1/2/z/<null>/pk2` here? If not, could you call that out here?